### PR TITLE
Add pprof profiling support to all binaries.

### PR DESCRIPTION
**Pull request by naemono**
## What is this change?

This adds the ability to expose the [pprof](https://pkg.go.dev/net/http/pprof) profiler for all of the topolvm binaries.

### Background

We are seeing a bit of memory growth over time with lvmd, and wanted to run some profiling to see if there's some sort of memory leak, but don't have that ability. This would add that functionality.

### CODE REVIEW

**Comment by naemono** [comment]
## Testing:

Tested locally  in minikube

```
❯ helm upgrade -i -n topolvm-system topolvm ../../charts/topolvm -f manifests/values/local.yaml
❯ kc get pod -n topolvm-system
NAME                                               READY   STATUS    RESTARTS      AGE
topolvm-cert-manager-657b7864b7-hhw2k              1/1     Running   0             9m13s
topolvm-cert-manager-cainjector-57fbb46b78-7k6nq   1/1     Running   0             9m13s
topolvm-cert-manager-webhook-85bff86bcc-zcj7w      1/1     Running   0             9m13s
topolvm-controller-664c66c87c-v46j4                5/5     Running   0             34s
topolvm-lvmd-0-t4qh8                               1/1     Running   0             33s
topolvm-node-58smh                                 3/3     Running   1 (31s ago)   33s

❯ diff manifests/values/base.yaml manifests/values/local.yaml 
2,4c2,4
<   repository: topolvm
<   tag: dev
<   pullPolicy: Never
---
>   repository: docker.io/personal/topolvm-with-sidecar
>   tag: devel
>   pullPolicy: Always
11a12,13
>   profiling:
>     port: 6060
46a49,50
>   profiling:
>     port: 6060
49a54,55
>   profiling:
>     port: 6060

```

File: internal/profiling/server.go
Comment (jakobmoellerdev) [inline internal/profiling/server.go:6]: please use a concrete import here, not the side effects based one, as this will automatically start pprof on the default handler (https://go.dev/src/net/http/pprof/pprof.go look into the init function). if you want to do it manually:

https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/manager/internal.go#L312-L327 see here for an example

File: cmd/lvmd/app/root.go
Comment (jakobmoellerdev) [inline cmd/lvmd/app/root.go:113]: This goroutine leaks and is not controlled by anything. Its preferrable to add it via controller-runtimes manager.Add()

**Review by jakobmoellerdev — CHANGES_REQUESTED** [review]
I think this PR in general is a great idea, thank you for your submission. However I have a few concerns:

1. Why do you not choose to use the existing pprof implementation at https://github.com/topolvm/topolvm/blob/ebfa7d02ad403a3a60fa9031a2d1796959e0dba8/cmd/topolvm-controller/app/run.go#L74-L89 via [PprofBindAddress](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.4/pkg/manager/manager.go#L237-L242)?
2. The goroutine you start uses a generic http handler and also is not tracked via the manager, making shutdown leak the goroutine

Last but not least, it would be great if you could share your memory observations over time. The leak findings are insanely useful and we will do our best to remediate

**Comment by jakobmoellerdev** [comment]
Another thing of note. If you use the images from topolvm directly, please note that we strip debugging information: https://github.com/topolvm/topolvm/blob/2096240b0b18dcfb4e4656259c5619176a2b102e/Makefile#L163 

It would be cool if we could add pprof when debugging symbols are enabled, but Ill leave it to you if you want to implement that or not

File: cmd/lvmd/app/root.go
Comment (toshipp) [reply cmd/lvmd/app/root.go:113]: Agreed.

**Comment by naemono** [comment]
> 1. Why do you not choose to use the existing pprof implementation at

I wasn't aware of this implementation in `controller-runtime`. I will migrate this to use that instead. Thanks for the pointer.

File: cmd/lvmd/app/root.go
Comment (naemono) [reply cmd/lvmd/app/root.go:113]: I'll get these non-controller-runtime implementations moved over to something that is able to be cancelled properly via a context/termination signal.. :+1:

**Comment by naemono** [comment]
> Last but not least, it would be great if you could share your memory observations over time. The leak findings are insanely useful and we will do our best to remediate

I will share more information when I have the details. At this time I don't have a ton to go on, just observation of memory growth over time that seems specific to the lvmd process.

**Comment by naemono** [comment]
Ok, the current changes use the `controller-runtime`'s pprof integration when available, otherwise it handled shutdown gracefully via `context` and doesn't leak goroutines. 

As for the note/question about stripping debugging information, were you suggesting detecting the debugging symbol's existence at runtime, and only serving the pprof endpoint when the debugging symbols exist? In my testing, even without the debugging symbols the pprof endpoint is rather useful. Let me know if you want some changes in this area. Thanks much for the quick review!

File: cmd/lvmd/app/config.go
Comment (jakobmoellerdev) [inline cmd/lvmd/app/config.go:27]: this is the zero value and you can just leave it out of the definition of the struct

File: cmd/lvmd/app/config.go
Comment (jakobmoellerdev) [inline cmd/lvmd/app/config.go:22]: how about changing this to a bind address instead so that people can not only control the port but also the network interface?

File: internal/profiling/server.go
Comment (jakobmoellerdev) [inline internal/profiling/server.go:13]: if you forward this entire address as argument instead of the port it allows people to use a different (e.g. non exposed) network for debugging in production environments without security risk.

File: cmd/topolvm-node/app/run.go
Comment (jakobmoellerdev) [inline cmd/topolvm-node/app/run.go:78]: similarly here, just forward the entire pprof bind address as an arg :)

**Review by jakobmoellerdev — CHANGES_REQUESTED** [review]
Mostly LGTM, just want to make sure we expose the entire bind address, not just the port. Thanks for taking care!

File: cmd/lvmd/app/root.go
Comment (toshipp) [inline cmd/lvmd/app/root.go:117]: The assignment should be out of the go statement because the shutdown code might access `pprofServer` before this.

File: cmd/topolvm-scheduler/app/root.go
Comment (toshipp) [inline cmd/topolvm-scheduler/app/root.go:109]: Also move this.

**Review by toshipp — CHANGES_REQUESTED** [review]

File: cmd/lvmd/app/config.go
Comment (naemono) [reply cmd/lvmd/app/config.go]: All implementations have been moved to using a bind address instead.

File: cmd/lvmd/app/root.go
Comment (jakobmoellerdev) [inline cmd/lvmd/app/root.go:132]: This is the wrong background context and also the shutdown should be bound to the base context

Comment (toshipp) [reply cmd/lvmd/app/root.go:132]: Since the `ctx` has been closed, we can not use it, IMO.

File: charts/topolvm/templates/controller/deployment.yaml
Comment (toshipp) [inline charts/topolvm/templates/controller/deployment.yaml:67]: We should update this. Must check other places.

**Review by toshipp — CHANGES_REQUESTED** [review]

File: cmd/lvmd/app/root.go
Comment (toshipp) [reply cmd/lvmd/app/root.go:132]: Hmm, we already call `grpcServer.GracefulStop`, the `subMain` may exit before shutting down pprofServer. It would be better to care about this case or close the server immediately.

File: charts/topolvm/templates/controller/deployment.yaml
Comment (naemono) [reply charts/topolvm/templates/controller/deployment.yaml:67]: :man_facepalming: I didn't test this after the last update or I would have caught this. I'll get this fixed. Thanks.

Comment (naemono) [reply charts/topolvm/templates/controller/deployment.yaml]: Ok, these have all been updated, and tested with the helm chart in minikube.

File: cmd/lvmd/app/root.go
Comment (naemono) [reply cmd/lvmd/app/root.go]: @jakobmoellerdev I believe I understand what you're asking for now, and have made the changes. Lmk, thanks.

**Review by jakobmoellerdev — APPROVED** [review]
LGTM, thanks so much for the submission again! :)

**Review by toshipp — CHANGES_REQUESTED** [review]

File: charts/topolvm/values.yaml
Comment (naemono) [reply charts/topolvm/values.yaml:114]: You are correct, no longer true. removed.

File: cmd/lvmd/app/config.go
Comment (naemono) [reply cmd/lvmd/app/config.go]: I was unaware of this option. I have removed this, and added a conditional cli flag.

File: charts/topolvm/values.yaml
Comment (toshipp) [reply charts/topolvm/values.yaml:114]: Could you please update the comments in this file?

**Review by toshipp — CHANGES_REQUESTED** [review]
mostly good. I'll approve this after you update values.yaml.
Also, these commits can be squashed, can you do that?

**Comment by naemono** [comment]
> mostly good. I'll approve this after you update values.yaml. Also, these commits can be squashed, can you do that?

Values file updated, and I'm working to squash these commits. :+1:

**Review by jakobmoellerdev — DISMISSED** [review]

**Comment by naemono** [comment]
@jakobmoellerdev It appears the helm readme hadn't been updated with `helm-docs`. I've run this update and will check back the ci tasks.

**Review by jakobmoellerdev — APPROVED** [review]

**Review by toshipp — COMMENTED** [review]
@naemono 
Please squash two commits.

> Add pprofl profiling support to all binaries.

I suspect `pprofl` is a typo of `pprof`. If true, please fix the commit message.

**Review by toshipp — APPROVED** [review]
LGTM. Thank you for your contribution!

**Comment by llamerada-jp** [comment]
Hi @naemono ,

> We are seeing a bit of memory growth over time with lvmd, and wanted to run some profiling to see if there's some sort of memory leak, but don't have that ability. This would add that functionality.

FYI, our team also found lvmd memory growth. I don't know if this is the same issue as your report.

The issue we found was that as we repeatedly created and deleted PVs in TopoLVM, the metrics `container_memory_usage_bytes{namespace="topolvm-system", container="lvmd"}` kept increasing. After investigation, this was not lvmd, but rather an increase in the lvm cache. Therefore, the following metrics were not increasing
`container_memory_usage_bytes{namespace="topolvm-system", container="lvmd"} - container_memory_cache{namespace="topolvm-system", container="lvmd"}`

Once we've reported this issue to LVM mailing list. However, we don't have enought time to dig into this problem for now.
https://lore.kernel.org/linux-lvm/ZjO_tW3_5ZNhnHFm@redhat.com/T/

**Pull request merged**
